src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs

```diff
@@ -798,10 +787,26 @@ public override unsafe void WriteDecimalText(decimal d)
             byte[] buffer = GetTextNodeBuffer(1 + sizeof(decimal), out offset);
             byte* bytes = (byte*)&d;
             buffer[offset++] = (byte)XmlBinaryNodeType.DecimalText;
-            for (int i = 0; i < sizeof(decimal); i++)
+            if (BitConverter.IsLittleEndian)
             {
-                buffer[offset++] = bytes[i];
+                for (int i = 0; i < sizeof(decimal); i++)
+                {
+                    buffer[offset++] = bytes[i];
+                }
+            }
+            else
+            {
+                Span<int> bits = stackalloc int[4];
+                decimal.TryGetBits(d, bits, out int intsWritten);
+                Debug.Assert(intsWritten == 4);
+
+                Span<byte> span = buffer.AsSpan(offset, sizeof(decimal));
+                BinaryPrimitives.WriteInt32LittleEndian(span, bits[3]);
+                BinaryPrimitives.WriteInt32LittleEndian(span.Slice(4), bits[2]);
```

Comment:
I'm not sure if this is correct. A decimal value is a 96 bit integer stored in little endian order with a final 32 bits holding some flags (stores a ^10 exponent divider and a sign bit). The order to be written to the span should be:  
[low 32 bits] [mid 32 bits] [high 32 bits] [flags]  
This looks like it writes out:
[flags] [high 32 bits] [low 32 bits] [mid 32 bits]  
While each of the 32bit numbers should be written in little endian so need their byte orders switched as you've done, the TryGetBits method doesn't write each of the components out to the passed in span in an endian specific way. The order of ints in the span should remain the same.  
Or am I missing something here?

Reply:
decimal is a struct with two Int32 fields and one Int64 field. The [declared order](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Decimal.cs,126) is flags, high32, low64. 

So when the little-endian implementation iterates over 4 integer pointers here, it writes out flags, high32, bottom-of-low64, top-of-low64 if I'm not mistaken, because the "low" is stored as Int64 and that's the order it would come out on a little-endian machine? Personally, I think it would make more sense to follow the order of TryGetBits if this was a greenfield implementation... but the goal here is to match on the wire what gets written by the little-endian implementation.

Or maybe I'm missing something? This is why I asked for extra eyes. ;)

Or that's how I read this code

Reply:
Yes, I agree with @StephenMolloy - we need to match the existing wire format, which matches the in-memory format of the `decimal` type on a little-endian machine, which is "[flags] [high 32] [low 32] [mid 32]".